### PR TITLE
Add id prefix, fact-scope unique constraint, and api audits

### PR DESCRIPTION
**Pull request by zhouzhuojie**
- Add id prefix
- Add scope-fact unique constraint check on non-empty scopes
- Refactor default mixin with ID() and deleted_at
- Add API Audits endpoints
- Migrate `make ent` and `make swag` to use docker-compose instead of local one to avoid breaking changes
- Further refactor the `apimodel.HTTPError` and `repo.Error`

### CODE REVIEW

File: pkg/repo/repo_ent_impl.go [hunk 1]
```diff
@@ -143,6 +143,35 @@ func (e *entImpl) AddPolicy(params ...interface{}) (bool, error) {
...
+		fact.Domain(opt.Domain),
+	).Exist(ctx)
+
+	// already exists, we shouldn't create a new fact with the same combination of [scope, fact_type, hashed_value]
+	// return validation error
+	if err != nil {
+		return nil, err
+	}
+	if exists {
+		return nil, fmt.Errorf("hashed_value already exists")
```
Comment (zhouzhuojie) [inline pkg/repo/repo_ent_impl.go:172]: @jperichon I may need to rebase and use the repo level of not validated error once your PR merged.

File: pkg/ent/schema/facttype.go [hunk 2]
```diff
@@ -15,6 +15,7 @@ type FactType struct {
 // Fields of the FactType.
 func (FactType) Fields() []ent.Field {
 	return []ent.Field{
+		ID("facttype"),
```
Comment (jperichon) [inline pkg/ent/schema/facttype.go:18]: fact_type?

Comment (zhouzhuojie) [reply pkg/ent/schema/facttype.go]: updated

File: pkg/repo/repo_ent_impl.go [hunk 3]
```diff
@@ -136,37 +138,84 @@ func (e *entImpl) Close() {
...
-	}
-	if _, ok := err.(govalidator.Errors); ok {
-		return NewValidationError(err, "Validation error")
-	}
-	if _, ok := err.(validator.ValidationErrors); ok {
-		return NewValidationError(err, "Validation error")
+		return ValidationError{Err: err, Message: "Validation error"}
 	}
 	if ent.IsConstraintError(err) {
 		if strings.Contains(err.Error(), "UNIQUE constraint failed: facts.hashed_value, facts.scope_facts, facts.fact_type_facts") {
```
Comment (zhouzhuojie) [inline pkg/repo/repo_ent_impl.go:153]: @jperichon what's the error message from postgres? I mean if we get `ent.IsConstraintError(err)`, maybe it's safe to assume it's a `ValidationError` without actually checking the string contains - at least for fact creation.

Comment (jperichon) [reply pkg/repo/repo_ent_impl.go:153]: From postgres:

```
"ent: constraint failed: insert node to table \"facts\": pq: duplicate key value violates unique constraint \"fact_hashed_value_scope_facts_fact_type_facts\"""
```

For sqlite:
```
"ent: constraint failed: insert node to table \"facts\": UNIQUE constraint failed: facts.hashed_value, facts.scope_facts, facts.fact_type_facts"
```

I made an attempt for fix at: https://github.com/open-privacy/opv/pull/33

File: pkg/repo/repo.go [hunk 4]
```diff
@@ -35,7 +35,7 @@ type Repo interface {
 	GrantRepo
 	APIAuditRepo
 
-	HandleError(ctx context.Context, err error) error
+	HandleError(err error) error
```
Comment (jperichon) [inline pkg/repo/repo.go:38]: 💯

File: cmd/controlplane/docs/swagger.yaml [hunk 5]
```diff
@@ -20,25 +39,70 @@ definitions:
...
+        type: object
+    type: object
 host: localhost
 info:
   contact: {}
   description: Open Privacy Vault Control Plane API.
   title: Open Privacy Vault Control Plane API
   version: "1.0"
 paths:
+  /api_audits:
```
Comment (jperichon) [inline cmd/controlplane/docs/swagger.yaml:59]: does this need pagination?

File: pkg/dataplane/dataplane.go [hunk 6]
```diff
@@ -73,11 +71,11 @@ func (dp *DataPlane) prepareEcho() {
 	}
 
 	apiv1 := e.Group("/api/v1")
+	apiv1.Use(dp.middlewareAPIAudit())
```
Comment (jperichon) [inline pkg/dataplane/dataplane.go:74]: should we also use that middleware in control plane?

File: pkg/apimodel/httperror.go [hunk 7]
```diff
@@ -1,56 +1,35 @@
...
+		// TODO or we can sanitize the error message with default status text
+		// return echo.NewHTTPError(e.Code)
+	case repo.NotFoundError:
+		return echo.ErrNotFound
+	case repo.ValidationError:
+		return echo.NewHTTPError(http.StatusBadRequest, err.Error())
+	case repo.UnauthorizedError:
+		return echo.ErrUnauthorized
+	case govalidator.Error, govalidator.Errors, validator.ValidationErrors:
+		return echo.NewHTTPError(http.StatusBadRequest, "Validation Error")
```
[the remark this answers is not in the document]
Comment (zhouzhuojie) [reply pkg/apimodel/httperror.go:30]: it flatterns to 
```
Status Code
4xx, 5xx

Body
{ "message": "" }
```

I actually learned something from https://github.com/labstack/echo/blob/master/echo.go#L108-L112

It has a good design of not returning "Code" and "Error" when serializing the JSON, the `json:"-"` part, and the `Error()` interface takes care of the `DEBUG` setting, which means if debug is enabled, the error message can also include the actual underlying error.

File: pkg/config/env.go [hunk 8]
```diff
@@ -18,7 +18,7 @@ var ENV = struct {
 	DBDriver                  string        `env:"OPV_DB_DRIVER" envDefault:"sqlite3"`
 	DBSetupRetryAttempts      uint          `env:"OPV_DB_SETUP_RETRY_ATTEMPTS" envDefault:"9"`
 	DBSetupRetryDelay         time.Duration `env:"OPV_DB_SETUP_RETRY_DELAY" envDefault:"100ms"`
-	DBConnectionStr           string        `env:"OPV_DB_CONNECTION_STR" envDefault:"_opv.sqlite?cache=shared&_fk=1"`
+	DBConnectionStr           string        `env:"OPV_DB_CONNECTION_STR" envDefault:"file:memdb1?mode=memory&cache=shared&_fk=1"`
```
Comment (zhouzhuojie) [inline pkg/config/env.go:21]: made a change to the default in-memory sqlite db, better for testing, and `make run` will have its own OPV_DB_CONNECTION_STR pointing to a local file.

**Pull request merged**